File: android/app/src/main/kotlin/dev/omnesis/android/ui/home/MenuRevealContainer.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.home

import android.os.Build
import android.view.HapticFeedbackConstants
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.PointerEventPass
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.input.pointer.util.VelocityTracker
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import dev.omnesis.android.designsystem.theme.OmTheme
import kotlin.math.abs
import kotlin.math.roundToInt
import kotlinx.coroutines.launch

/**
 * Presents the main menu *beneath* the app, and slides the app aside to reveal it.
 *
 * Moving the app, rather than covering it, keeps it the object under the user's hand: it
 * steps aside, corners rounding as it lifts, and stays visible on the trailing edge as
 * somewhere to go back to. That also frees the gesture to work across the whole screen,
 * since there is no panel that has to be dragged in from an edge.
 *
 * Layering, bottom to top: [menu], then [content] offset by `progress` of the travel,
 * washed by a scrim and cut to a corner radius that both scale with the same progress.
 *
 * ## Losing the drag on purpose
 *
 * The reveal shares the rightward direction with things nearer the finger — a code block
 * that scrolls sideways, a list row moving its swipe actions. The pointer handler runs on
 * [PointerEventPass.Main], which reaches a parent only *after* its children, and refuses
 * any drag a child has already consumed. That is what keeps the two from firing at once.
 * The iOS app needed a UIKit recogniser and `shouldRequireFailureOf` to say the same
 * thing; Compose says it with a pass and a consumption check.
 */
@Composable
fun MenuRevealContainer(
    isOpen: Boolean,
    onOpenChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
    /**
     * Freezes the reveal at a fixed point of its travel (0 closed, 1 open), ignoring
     * [isOpen] and any drag. Screenshot goldens use it to render a mid-gesture frame,
     * which is otherwise unreachable in a still — and to render the open state at all,
     * since the settle animation does not advance under a Robolectric render.
     */
    progressOverride: Float? = null,
    /** Test seam for the committed release; null performs the platform feedback. */
    onCommittedSwipeHaptic: (() -> Unit)? = null,
    menu: @Composable () -> Unit,
    content: @Composable () -> Unit,
) {
    val colors = OmTheme.colors
    val density = LocalDensity.current
    val scope = rememberCoroutineScope()
    val view = LocalView.current
    val currentCommittedSwipeHaptic by rememberUpdatedState(onCommittedSwipeHaptic)

    // The menu slides the app aside, so whatever was being typed into is no longer reachable.
    // Done here rather than at each call site because a tap, a drag and a deep link all arrive
    // as the same `isOpen`.
    val keyboard = LocalSoftwareKeyboardController.current
    val focusManager = LocalFocusManager.current
    LaunchedEffect(isOpen) {
        if (isOpen) {
            focusManager.clearFocus(force = true)
            keyboard?.hide()
        }
    }

    var widthPx by remember { mutableFloatStateOf(0f) }
    val travelPx = MenuReveal.menuWidthPx(widthPx, density.density)
    // The app's distance from rest, in pixels. An `Animatable` so a released drag carries
    // on from wherever the finger left it, rather than snapping back and re-animating.
    val offset = remember { Animatable(0f) }
    var dragging by remember { mutableStateOf(false) }
    var dragOffsetPx by remember { mutableFloatStateOf(0f) }

    // Keeps the app in step with `isOpen` when it changes from anywhere but a drag — a
    // toolbar tap, a menu row, a back press.
    LaunchedEffect(isOpen, travelPx, progressOverride) {
        if (travelPx <= 0f || progressOverride != null) return@LaunchedEffect
        val target = if (isOpen) travelPx else 0f
        if (abs(offset.value - target) > 0.5f) offset.animateTo(target, tween(220))
    }

    val liveOffsetPx = if (dragging) dragOffsetPx else offset.value
    val progress = progressOverride?.coerceIn(0f, 1f)
        ?: MenuReveal.progress(liveOffsetPx, travelPx)
    val offsetPx = if (progressOverride != null) travelPx * progress else liveOffsetPx

    Box(
        modifier
            .fillMaxSize()
            .background(colors.bgDrawer)
            .onSizeChanged { widthPx = it.width.toFloat() }
            .pointerInput(isOpen, travelPx, progressOverride) {
                if (travelPx <= 0f || progressOverride != null) return@pointerInput
                awaitEachGesture {
                    // Main, not Initial: children see the event first, so anything nearer
                    // the finger has already had its chance to claim the drag.
                    val down = awaitFirstDown(requireUnconsumed = false, pass = PointerEventPass.Main)
                    val startX = down.position.x
                    val tracker = VelocityTracker()
                    tracker.addPosition(down.uptimeMillis, down.position)
                    // Where the app rests for this touch. The drag is applied as an
                    // absolute offset from here rather than an increment, so the settles
                    // launched below cannot land out of order.
                    val restingPx = if (isOpen) travelPx else 0f
                    var mode = MenuReveal.DragMode.Undetermined
                    var settling = false

                    try {
                        while (true) {
                            val event = awaitPointerEvent(PointerEventPass.Main)
                            val change = event.changes.firstOrNull { it.id == down.id } ?: break
                            tracker.addPosition(change.uptimeMillis, change.position)
                            val translation = change.position - down.position
                            val velocity = tracker.calculateVelocity()
                            mode = MenuReveal.classifyDrag(
                                current = mode,
                                startX = startX / density.density,
                                totalDx = translation.x / density.density,
                                totalDy = translation.y / density.density,
                                velocityX = velocity.x / density.density,
                                velocityY = velocity.y / density.density,
                                width = widthPx / density.density,
                                isOpen = isOpen,
                                consumedByChild = change.isConsumed,
                            )
                            if (!change.pressed) {
                                if (mode == MenuReveal.DragMode.Horizontal) {
                                    settling = true
                                    val releasedOffset = MenuReveal.clampOffset(restingPx + translation.x, travelPx)
                                    val open = MenuReveal.shouldOpen(
                                        velocityDpPerSec = velocity.x / density.density,
                                        dragDistancePx = translation.x,
                                        travelPx = travelPx,
                                        isOpen = isOpen,
                                    )
                                    if (MenuReveal.committedStateChanged(isOpen, open)) {
                                        if (currentCommittedSwipeHaptic != null) {
                                            currentCommittedSwipeHaptic?.invoke()
                                        } else {
                                            val feedback = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                                                HapticFeedbackConstants.GESTURE_END
                                            } else {
                                                HapticFeedbackConstants.CLOCK_TICK
                                            }
                                            view.performHapticFeedback(feedback)
                                        }
                                    }
                                    scope.launch {
                                        offset.snapTo(releasedOffset)
                                        dragging = false
                                        offset.animateTo(
                                            if (open) travelPx else 0f,
                                            spring(dampingRatio = 0.9f, stiffness = 400f),
                                        )
                                        onOpenChange(open)
                                    }
                                }
                                break
                            }
                            if (mode == MenuReveal.DragMode.Horizontal) {
                                dragging = true
                                dragOffsetPx = MenuReveal.clampOffset(restingPx + translation.x, travelPx)
                                change.consume()
                            }
                        }
                    } finally {
                        // Cancellation, pointer loss, or a pointerInput key restart has no release
                        // sample to settle. Drop the transient offset so it cannot pin the app.
                        if (!settling) {
                            dragging = false
                        }
                    }
                }
            },
    ) {
        // The menu is only ever as wide as the app moves — past that it is behind the app
        // and could never be seen. Laying it out to the full screen instead would run its
        // rows, and the Settings button at its trailing edge, underneath the standing
        // strip where they are permanently unreachable.
        Box(
            Modifier
                .align(Alignment.CenterStart)
                // Width BEFORE height: `fillMaxSize` would pin the width to the incoming
                // maximum and a later `width` could not shrink it — which lays the menu
                // out full-screen and puts its trailing Settings button underneath the
                // standing strip, where it can never be reached.
                .then(
                    if (travelPx > 0f) Modifier.width(with(density) { travelPx.toDp() }) else Modifier,
                )
                .fillMaxHeight()
                // Behind the app rather than off-screen, so a screen reader would
                // otherwise reach it straight through whatever is on top.
                .then(if (progress < 1f) Modifier.clearAndSetSemantics {} else Modifier),
        ) {
            menu()
        }

        AppLayer(
            progress = progress,
            offsetPx = offsetPx,
            onTapToClose = {
                scope.launch {
                    offset.animateTo(0f, tween(220))
                    onOpenChange(false)
                }
            },
            content = content,
        )
    }
}

/**
 * The app itself: offset, cut to a rounded corner that grows as it lifts, washed by a
 * scrim, and edged with a hairline.
 *
 * In dark mode the hairline, not the scrim, is what separates the app from the menu — the
 * two surfaces are only a few levels apart and `border` is lighter than both.
 */
@Composable
private fun AppLayer(
    progress: Float,
    offsetPx: Float,
    onTapToClose: () -> Unit,
    content: @Composable () -> Unit,
) {
    val colors = OmTheme.colors
    val shape = RoundedCornerShape((MenuReveal.SCREEN_CORNER_RADIUS_DP * progress).dp)
    Box(
        Modifier
            .fillMaxSize()
            .offset { IntOffset(offsetPx.roundToInt(), 0) }
            .shadow(elevation = (24f * progress).dp, shape = shape, clip = false)
            .clip(shape)
            .background(colors.bgPrimary)
            .border(1.dp, colors.border.copy(alpha = progress), shape),
    ) {
        Box(
            Modifier
                .fillMaxSize()
                .then(if (progress > 0f) Modifier.clearAndSetSemantics {} else Modifier),
        ) {
            content()
        }
        if (progress > 0f) {
            // The wash, and — once the app has fully stepped aside — the way back. It also
            // swallows touches meant for the app underneath, which is inert while it is
            // standing aside.
            Box(
                Modifier
                    .fillMaxSize()
                    .background(Color.Black.copy(alpha = MenuReveal.SCRIM_OPACITY * progress))
                    .pointerInput(progress >= 1f) {
                        if (progress >= 1f) {
                            detectTapGestures { onTapToClose() }
                        } else {
                            // Mid-travel: absorb touches without acting on them, so a
                            // stray tap never lands on the app that is on its way out.
                            awaitEachGesture {
                                awaitFirstDown(requireUnconsumed = false, pass = PointerEventPass.Main)
                            }
                        }
                    },
            )
        }
    }
}

File: ios/Sources/Omnesis/UI/MenuRevealContainer.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(SwiftUI) && canImport(UIKit)
import SwiftUI
import UIKit

@available(iOS 17.0, *)
extension MenuReveal {
    /// Every non-interactive open/close — a toolbar tap, a menu row, a
    /// deep-link that dismisses the menu on its way somewhere else.
    static let openCloseAnimation: Animation = .easeOut(duration: 0.22)

    /// Settle for a *released drag*, launched at the speed the finger left.
    static func releaseAnimation(initialVelocity: Double) -> Animation {
        .interpolatingSpring(mass: 1.0, stiffness: 240, damping: 28, initialVelocity: initialVelocity)
    }
}

/// Frames of the regions that own horizontal drags themselves.
///
/// A reference box rather than `@State`: the frames change on every frame of
/// any vertical scroll that carries one of these regions, and of the reveal
/// itself, since they are resolved in the container's coordinate space. Held
/// in `@State` that would invalidate the container's body — and with it the
/// whole app — once per frame. The gesture is the only reader, and it reads at
/// touch-down, so nothing needs to redraw when they move.
private final class MenuRevealExclusions {
    var frames: [CGRect] = []
}

/// Regions that own horizontal drags themselves, reported up to the container.
private struct MenuRevealExclusionKey: PreferenceKey {
    static var defaultValue: [CGRect] {
        []
    }

    static func reduce(value: inout [CGRect], nextValue: () -> [CGRect]) {
        value.append(contentsOf: nextValue())
    }
}

@available(iOS 17.0, *)
extension View {
    /// Marks a subtree that scrolls horizontally under the user's finger — a
    /// code block, a wide table, a rail of cards, a panel that closes on a
    /// rightward drag — so a drag starting inside it moves that content and
    /// nothing else.
    ///
    /// Without this both gestures fire: the region scrolls *and* the menu
    /// comes out from under the app, which is where a reveal spanning most of
    /// the screen is genuinely ambiguous. Declaring it at the scrolling view
    /// keeps the knowledge where it belongs — the container holds no list of
    /// special cases, and a new horizontal scroller opts in with one modifier.
    ///
    /// Pass `isActive: false` whenever the region is not currently taking
    /// touches — a panel parked off-screen, a rail that is hidden. The frame
    /// reported here is the **layout** frame, and `.offset` is a draw-time
    /// transform that does not move it: a panel slid out of sight still
    /// reports the frame it would occupy, and would go on excluding that whole
    /// area. Gate on the same condition that governs the region's hit-testing
    /// and the two can never disagree.
    ///
    /// Only meaningful inside a `MenuRevealContainer`; harmless anywhere else,
    /// and inert inside a sheet, whose content reports no preferences to the
    /// presenter it covers.
    func menuRevealExcluded(isActive: Bool = true) -> some View {
        background {
            GeometryReader { geo in
                Color.clear.preference(
                    key: MenuRevealExclusionKey.self,
                    value: isActive ? [geo.frame(in: .named(MenuReveal.coordinateSpace))] : []
                )
            }
        }
    }
}

/// Presents the main menu *beneath* the app, and slides the app aside to
/// reveal it.
///
/// Moving the app, rather than covering it, keeps it the object under the
/// user's hand: it steps aside, corners rounding as it lifts, and stays
/// visible on the trailing edge as somewhere to go back to. That also frees
/// the gesture to work across the whole screen, since there is no panel that
/// has to be dragged in from an edge.
///
/// Layering, bottom to top: `menu`, then `content` offset by `progress` of the
/// travel, washed by a scrim and cut to a corner radius that both scale with
/// the same `progress`. One drag gesture on the container drives that progress
/// in both directions; `MenuReveal` holds every value it is tuned by.
///
/// The container never lays out either layer differently from how they would
/// lay out alone — `offset` and `mask` are draw-time only — so the app keeps
/// its own safe areas, and the width the offset is measured against is read
/// from a background probe rather than a wrapping `GeometryReader`.
@available(iOS 17.0, *)
struct MenuRevealContainer<Menu: View, Content: View>: View {
    @Environment(\.colorScheme) private var colorScheme
    @Binding var isOpen: Bool
    /// Freezes the reveal at a fixed point of its travel (0 closed, 1 open),
    /// ignoring `isOpen` and any drag. Previews and snapshot tests use it to
    /// render a mid-gesture frame, which is otherwise unreachable in a still;
    /// when set, the drag is inert so nothing can move behind a frozen render.
    let progressOverride: CGFloat?
    private let menu: Menu
    private let content: Content

    /// Container width, read from a background probe. Zero until first layout,
    /// which resolves to a closed menu — the state the app launches in anyway.
    @State private var width: CGFloat = 0
    /// Distance covered by the drag in progress, signed. Held in `@State`
    /// rather than `@GestureState` so it survives touch-up and can be animated
    /// back to zero in the same `withAnimation` block as `isOpen`: that joint
    /// settle is what carries the app on from wherever the finger left it,
    /// instead of snapping back and re-animating from the resting position.
    @State private var dragDistance: CGFloat = 0
    @State private var exclusions = MenuRevealExclusions()
    @State private var feedbackGenerator = UIImpactFeedbackGenerator(style: .soft)

    init(
        isOpen: Binding<Bool>,
        progressOverride: CGFloat? = nil,
        @ViewBuilder menu: () -> Menu,
        @ViewBuilder content: () -> Content
    ) {
        self._isOpen = isOpen
        self.progressOverride = progressOverride
        self.menu = menu()
        self.content = content()
    }

    /// Points the app travels between closed and fully open — which is also
    /// exactly the width it uncovers, hence the menu's own width.
    private var travel: CGFloat {
        MenuReveal.menuWidth(forContainerWidth: width)
    }

    private var restingDistance: CGFloat {
        isOpen ? travel : 0
    }

    private var progress: CGFloat {
        if let progressOverride { return MenuReveal.clamped(unit: progressOverride) }
        return MenuReveal.progress(
            restingDistance: restingDistance,
            dragDistance: dragDistance,
            travel: travel
        )
    }

    /// Width the menu lays out in. `nil` before first layout, which leaves it
    /// unconstrained for that one frame rather than collapsing it to nothing.
    private var menuWidth: CGFloat? {
        width > 0 ? travel : nil
    }

    var body: some View {
        ZStack(alignment: .leading) {
            // Painted edge to edge even though the menu is laid out narrower,
            // so no seam can show between the two at full travel.
            Theme.bgDrawer.ignoresSafeArea()
            menu
                // The menu is only ever as wide as the app moves — past that
                // it is behind the app and could never be seen. Laying it out
                // to the full screen instead would run its rows, and the
                // Settings button at its trailing edge, underneath the
                // standing strip where they are permanently unreachable.
                .frame(width: menuWidth, alignment: .leading)
                // Inert until the menu has fully arrived, so a finger that
                // lands on a row and then swipes the app back over it does not
                // also fire that row on release.
                .disabled(progress < 1)
                // Hidden from VoiceOver on the *committed* state, not on
                // `progress`: a pair of thresholds either side of a moving
                // value leaves mid-travel with both layers hidden and nothing
                // at all to read.
                .accessibilityHidden(!isOpen)
            contentLayer
        }
        .background {
            GeometryReader { geo in
                Color.clear
                    .onAppear { width = geo.size.width }
                    .onChange(of: geo.size.width) { _, newWidth in width = newWidth }
            }
        }
        .coordinateSpace(.named(MenuReveal.coordinateSpace))
        .onPreferenceChange(MenuRevealExclusionKey.self) { [exclusions] frames in
            exclusions.frames = frames
        }
        // A real recogniser, on the window, so the reveal can lose to the
        // gestures UIKit owns instead of firing alongside them. Laid in the
        // background: it hit-tests to nothing and only exists to own the
        // recogniser's lifetime.
        .background {
            MenuRevealPanGesture(
                shouldBegin: { start, translation, velocity in
                    let accepted = MenuReveal.mode(
                        velocity: velocity,
                        startLocation: start,
                        translation: translation,
                        width: width,
                        isOpen: isOpen,
                        exclusions: exclusions.frames
                    ) == .horizontal && progressOverride == nil
                    if accepted { feedbackGenerator.prepare() }
                    return accepted
                },
                onChanged: { translation in
                    dragDistance = MenuReveal.clampDragDistance(
                        translation.x,
                        restingDistance: restingDistance,
                        travel: travel
                    )
                },
                onEnded: { velocity in
                    guard travel > 0 else {
                        dragDistance = 0
                        return
                    }
                    settle(velocity: velocity.x)
                },
                onCancelled: {
                    withAnimation(MenuReveal.openCloseAnimation) { dragDistance = 0 }
                }
            )
        }
        // VoiceOver's two-finger scrub, and the equivalent on Switch Control:
        // without it the menu can be opened but only left by committing to a
        // destination, since the scrim is a tap target no assistive technology
        // announces and the toolbar button is inside the hidden app.
        .accessibilityAction(.escape) { close() }
    }

    private var contentLayer: some View {
        content
            // Cancels whatever the touch had started on the moment it becomes
            // a reveal, so a finger that landed on a row and then swiped does
            // not also fire that row on release.
            .disabled(progress > 0)
            // Applied before `.overlay` so the scrim, which is the way back,
            // stays both interactive and audible while the app is inert.
            .accessibilityHidden(isOpen)
            .overlay { scrim }
            // Masked rather than clipped, with the *mask* ignoring the safe
            // area. A clip would cut the app off at its safe-area frame and
            // let the menu show through behind the status bar; expanding the
            // app itself instead would push its navigation bar up under the
            // clock. Masking moves neither, and rounds the corners where the
            // screen's own corners are.
            .mask { revealShape.ignoresSafeArea() }
            // Drawn after the mask so it lands exactly on the cut edge. In
            // dark mode this hairline, not the scrim, is what separates the
            // app from the menu: the two surfaces are only a few levels
            // apart, and `Theme.border` is lighter than both.
            .overlay {
                revealShape
                    .strokeBorder(Theme.border.opacity(Double(progress)), lineWidth: 1)
                    .ignoresSafeArea()
                    .allowsHitTesting(false)
            }
            // The shadow is cast by a bare shape *behind* the app rather than
            // by the app itself — an identical edge, without rasterising a
            // live transcript offscreen on every frame of the drag. The fill
            // only exists to have something to cast it.
            .background {
                revealShape
                    .fill(Theme.bgPrimary)
                    .ignoresSafeArea()
                    .shadow(
                        color: .black.opacity(
                            MenuReveal.shadowOpacity(isLight: colorScheme == .light)
                                * Double(progress)
                        ),
                        radius: MenuReveal.shadowRadius(isLight: colorScheme == .light) * progress,
                        x: MenuReveal.shadowOffsetX * progress
                    )
            }
            .offset(x: travel * progress)
    }

    /// Corners round as the app lifts away, reaching the screen's own radius
    /// at full travel — so what stands on the trailing edge reads as the
    /// screen it came from, not a rectangle laid over one.
    private var revealShape: RoundedRectangle {
        RoundedRectangle(cornerRadius: MenuReveal.screenCornerRadius * progress, style: .continuous)
    }

    /// Wash over the departing app, and — once it has fully arrived — the tap
    /// target that brings it back.
    ///
    /// Always present, never conditional: a scrim inserted and removed by an
    /// `if` stays in the hierarchy for the length of its removal transition,
    /// still holding the hit-testing it had when the menu was open, and spends
    /// that time swallowing every tap in the app. Driven purely by `progress`
    /// it also cross-fades instead of popping.
    private var scrim: some View {
        Color.black
            .opacity(MenuReveal.scrimOpacity * Double(progress))
            // Reaches the status bar and the home indicator, which the app
            // paints behind and the overlay's own frame excludes.
            .ignoresSafeArea()
            .onTapGesture { close() }
            .accessibilityLabel("Close menu")
            .accessibilityAddTraits(.isButton)
            // Inert until the menu has fully arrived: transparent at rest, and
            // mid-drag it must never compete with the gesture placing it.
            .allowsHitTesting(progress >= 1)
            // Withdrawn from the accessibility tree by the same rule. Traits
            // survive `allowsHitTesting(false)`, so leaving this on would
            // spread a full-screen "Close menu" button across the app at rest:
            // VoiceOver would announce a control that is not there, and
            // everything beneath it — a list row's swipe actions among them —
            // would report as covered and refuse taps.
            .accessibilityHidden(progress < 1)
    }

    /// Carries the app on from where the finger left it, at the speed it left,
    /// to whichever end of the travel the drag committed to.
    private func settle(velocity: CGFloat) {
        let outcome = MenuReveal.releaseOutcome(
            velocity: velocity,
            dragDistance: dragDistance,
            travel: travel,
            isOpen: isOpen
        )
        let current = restingDistance + dragDistance
        let remaining = (outcome.shouldOpen ? travel : 0) - current
        if outcome.emitsFeedback {
            feedbackGenerator.impactOccurred(intensity: 0.7)
        }
        withAnimation(MenuReveal.releaseAnimation(
            initialVelocity: MenuReveal.initialVelocity(velocity: velocity, remaining: remaining)
        )) {
            isOpen = outcome.shouldOpen
            dragDistance = 0
        }
    }

    private func close() {
        withAnimation(MenuReveal.openCloseAnimation) { isOpen = false }
    }
}

#if DEBUG
/// Stand-in for a section view, so the container's own choreography can be
/// previewed and snapshotted without building a populated `AppStore`.
@available(iOS 17.0, *)
private struct MenuRevealPreviewContent: View {
    var body: some View {
        VStack(alignment: .leading, spacing: Theme.Spacing.md) {
            HStack {
                Image(systemName: "line.3.horizontal").foregroundStyle(Theme.textPrimary)
                Spacer()
                Text("Ask").font(.system(size: 17, weight: .semibold))
                    .foregroundStyle(Theme.textPrimary)
                Spacer()
                Image(systemName: "square.and.pencil").foregroundStyle(Theme.textPrimary)
            }
            .padding(.horizontal, Theme.Spacing.lg)
            ForEach(0 ..< 6) { index in
                RoundedRectangle(cornerRadius: Theme.Radius.large)
                    .fill(Theme.bgSecondary)
                    .frame(height: index.isMultiple(of: 2) ? 64 : 96)
                    .padding(.horizontal, Theme.Spacing.lg)
            }
            Spacer()
        }
        .padding(.top, Theme.Spacing.xl)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Theme.bgPrimary)
    }
}

@available(iOS 17.0, *)
private struct MenuRevealPreviewMenu: View {
    var body: some View {
        VStack(alignment: .leading, spacing: Theme.Spacing.lg) {
            Text("Omnesis").font(.system(size: 22, weight: .semibold))
                .foregroundStyle(Theme.textPrimary)
            ForEach(["Search", "Briefs", "Sources", "People"], id: \.self) { label in
                Text(label).font(.system(size: 18)).foregroundStyle(Theme.textPrimary)
            }
            Spacer()
        }
        .padding(.horizontal, Theme.Spacing.md)
        .padding(.top, 60)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.bgDrawer.ignoresSafeArea())
    }
}

@available(iOS 17.0, *)
#Preview("MenuReveal — open") {
    MenuRevealContainer(isOpen: .constant(true)) {
        MenuRevealPreviewMenu()
    } content: {
        MenuRevealPreviewContent()
    }
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("MenuReveal — mid-drag") {
    MenuRevealContainer(isOpen: .constant(false), progressOverride: 0.45) {
        MenuRevealPreviewMenu()
    } content: {
        MenuRevealPreviewContent()
    }
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("MenuReveal — mid-drag, light") {
    MenuRevealContainer(isOpen: .constant(false), progressOverride: 0.45) {
        MenuRevealPreviewMenu()
    } content: {
        MenuRevealPreviewContent()
    }
    .preferredColorScheme(.light)
}

// The first moments of the travel, where the corner radius is only a few
// points on an already-lifted shape — the frame most likely to read as an
// artefact rather than an intention.
@available(iOS 17.0, *)
#Preview("MenuReveal — barely moved") {
    MenuRevealContainer(isOpen: .constant(false), progressOverride: 0.08) {
        MenuRevealPreviewMenu()
    } content: {
        MenuRevealPreviewContent()
    }
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("MenuReveal — closed") {
    MenuRevealContainer(isOpen: .constant(false)) {
        MenuRevealPreviewMenu()
    } content: {
        MenuRevealPreviewContent()
    }
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("MenuReveal — open, light") {
    MenuRevealContainer(isOpen: .constant(true)) {
        MenuRevealPreviewMenu()
    } content: {
        MenuRevealPreviewContent()
    }
    .preferredColorScheme(.light)
}

// The real composition — the actual menu under the actual Agent surface —
// rather than the stand-ins above.
@available(iOS 17.0, *)
#Preview("MenuReveal — real menu over Agent") {
    @Previewable @State var isOpen = true
    return MenuRevealContainer(isOpen: $isOpen) {
        MainMenuDrawer(isOpen: $isOpen, selection: .constant(.agent), onOpenSettings: {})
    } content: {
        AgentView(menuOpen: $isOpen)
    }
    .environment(AppStore.preview(agentPreview: PreviewMocks.agentConversationsRich))
    .preferredColorScheme(.dark)
}
#endif
#endif
